smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -284,70 +266,133 @@ private void generateOperationSerializer(
+                        output, binding.getMember(), target);
+                writer.write("headers['$L'] = $L;", binding.getLocationName(), headerValue);
+            });
+        }
+
+        // Handle assembling prefix headers.
+        for (HttpBinding binding : bindingIndex.getRequestBindings(operation, Location.PREFIX_HEADERS)) {
+            String memberName = symbolProvider.toMemberName(binding.getMember());
+            writer.openBlock("if (input.$L !== undefined) {", "}", memberName, () -> {
+                Shape target = index.getShape(binding.getMember().getTarget()).get();
+                // Iterate through each entry in the member.
+                writer.openBlock("Objects.keys(input.$L).forEach(suffix -> {", "});", memberName, () -> {
+                    String headerValue = getInputValue(context, binding.getLocation(),
+                            "input." + memberName + "[suffix]", output, binding.getMember(), target);
+                    // Append the suffix to the defined prefix and serialize the value in to that key.
+                    writer.write("headers['$L' + suffix] = $L;", binding.getLocationName(), headerValue);
+                });
+            });
+        }
+    }
+
+    private List<HttpBinding> writeRequestBody(
+            GenerationContext context,
+            OperationShape operation,
+            HttpBindingIndex bindingIndex
+    ) {
+        TypeScriptWriter writer = context.getWriter();
+        List<HttpBinding> documentBindings = bindingIndex.getRequestBindings(operation, Location.DOCUMENT);
+        documentBindings.sort(Comparator.comparing(HttpBinding::getMemberName));
+        List<HttpBinding> payloadBindings = bindingIndex.getRequestBindings(operation, Location.PAYLOAD);
+
+        if (!documentBindings.isEmpty()) {
+            // Write the default `body` property.
+            context.getWriter().write("let body: any = undefined;");
+            serializeDocument(context, operation, documentBindings);
+            return documentBindings;
+        }
+        if (!payloadBindings.isEmpty()) {
+            SymbolProvider symbolProvider = context.getSymbolProvider();
+            // There can only be one payload binding.
+            HttpBinding binding = payloadBindings.get(0);
+            String memberName = symbolProvider.toMemberName(binding.getMember());
+            Shape target = context.getModel().getShapeIndex().getShape(binding.getMember().getTarget()).get();
+            writer.write("let body: any = $L;", getInputValue(
+                    context, Location.PAYLOAD, "input." + memberName, operation, binding.getMember(), target));
+            return payloadBindings;
+        }
+
+        return ListUtils.of();
+    }
+
     protected String getInputValue(
             GenerationContext context,
             Location bindingType,
-            Shape shape,
+            String dataSource,
```

Comment:
Passing in the `dataSource` allows for a significant reduction in the amount of context and type checking to alter the location. In all cases, the calling context had the right information to generate this properly.

Reply:
What is dataSource?